src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackageDependencyType.cs

```diff
@@ -18,6 +18,11 @@ public enum PackageDependencyType
         /// <summary>
         /// 
         /// </summary>
-        Project = 2
+        Project = 2,
+
+        /// <summary>
+        /// Package is transitively available to the project instead of directly installed but is also enforced by the central package version file.
```

Comment:
Can you elaborate why this is necessary?
Can we do without it?

Reply:
I agree. I don't think it's important in a lock file to know it was pinned because of central management, or if it was a 'normal' transitive dependency. Honestly, I don't think the lock file should care about direct vs transitive, but I guess it's too late for that.

The lock file just makes sure what was restored this time was the same as what was restored last time. why a version was selected is not important.

Reply:
> Can you elaborate why this is necessary?
> Can we do without it?

The scenarios that i could not achieve without this are:

1. Evaluate if the lock file needs to be invalidated when a transitive dependency has its central version removed from the CPVM file. 
The comparison is made between a package.lock.file and a dgspec 
The lock file contains transitive dependencies.
The dgspec conatins only direct deps and central versions.  
It is not necessary as a transitive dependency to be in the central file. But it can be enforced by the pinning as well.

If an entry is removed from the central file it can remove the pinning. However having the lock file (packages.lock.json) not knowing that it has central-transitive dep affected i could not find a way of invalidating or not. 

2. A lock (packages.lock.json) exists with transitive dep A (1.0.0) . Central version adds a PackageA version (>1.0.0). The lock file should be invalidated in this case, because the pining of transitive dep could resolve to 2.0.0. 

I would be more than happy to remove this new field if we can achieve these scenarios otherwise.

Reply:
> The lock file just makes sure what was restored this time was the same as what was restored last time. why a version was selected is not important.

The lock file validation happens before the current restore. I suppose that this was done to avoid unnecessary restore for performance reasons. Before the restore there is only the dgspec that does not have information about possible transitive dependencies.

Reply:
I glanced over the [repeatable build spec](https://github.com/NuGet/Home/wiki/Enable-repeatable-package-restore-using-lock-file), which defined the lock file. I don't know how accurately the implementation of lock files is to this spec. But a few things jumped at out me.

> A lock file has the package dependency graph for the project/solution/repo that includes both the direct as well as transitive dependencies.

Unfortunately "package dependency graph" is not well defined. As a NuGet developer, I know that CPVM's `PackageVersion` is in the DGSpec, so does that mean it's part of the "package dependency graph"? However, as a user, I would say no. The project lock file defines the packages used by the project. CPVM `PackageVersion`s for packages not used in the current project is not a dependency of the current project and therefore doesn't belong.

> The lock file format should be such that it is: Concise

it is arguably not concise adding additional entries that are not essential to repeatable build scenario

> The integrity of the package SHA512 or equivalent should be also persisted along with the lock file.

What would the lock file contain for the SHA512 of a CPVM `PackageVersion` that is not used by the project?

For these reasons, I feel like the proposed changes of adding a new dependency type for CPVM violates some of the design goals of the repeatable build feature.

Reply:
> The lock file validation happens before the current restore. I suppose that this was done to avoid unnecessary restore for performance reasons. Before the restore there is only the dgspec that does not have information about possible transitive dependencies.

The repeatable build spec says that if your project uses floating versions, the version in the lock file will be used, unless NuGet is explicitly asked to re-evaluate the latest version. I believe this is the reason that the lock file is read before restore.

It may be that lock file out of sync checks used an optimisation/assumption that is no longer true with CPVM. One option is to modify the lock file so that it still holds. Is another possibility to move the lock file validation check to after restore?

Consider this non-CPVM scenario. What happens if `MyProject` depends on `PackageA` version 1.0.0, and `PackageA` version 1.0.0 has a dependency on `PackageB` version `1.0.0`, but `PackageB` version 1.0.0 does not exist on any configured source, so version 1.0.1 was restored instead?  The lock file should list `PackageB` version 1.0.1, correct?  Well, what if 1.0.1 is removed the the source (or the sources change), and 1.0.1 is no longer available, only 1.0.2 is?  We would still expect locked mode restore to fail, right? And we would expect non-locked mode restore to update the lock file with the new version numer and hash that was actually restored, right?

I don't know the lock file/repeatable build feature very well, but it seems to me that it should already be able to handle the same DG spec before restore resulting in different actual restore graphs. Or is there a bug in this non-cpvm scenario?  If the non-cpvm scenario works, can cpvm work in a similar way to avoid needed to "pollute" the lock file?

Reply:
@zivkan did the offline meeting yesterday helped with understanding these changes for CPVM?